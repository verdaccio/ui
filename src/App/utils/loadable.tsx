import React, { lazy, Suspense } from 'react';
export default (importCallback: any) => {
  const TargetComponent = lazy(importCallback);
  return (props: any) => (
    <Suspense fallback={null}>
      <TargetComponent {...props} />
    </Suspense>
  );
};
