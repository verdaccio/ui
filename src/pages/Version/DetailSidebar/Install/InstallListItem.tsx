import styled from '@emotion/styled';
import React from 'react';
import { useTranslation } from 'react-i18next';

import Avatar from 'verdaccio-ui/components/Avatar';
import CopyToClipBoard from 'verdaccio-ui/components/CopyToClipBoard';
import ListItem from 'verdaccio-ui/components/ListItem';
import ListItemText from 'verdaccio-ui/components/ListItemText';

// logos of package managers
import npmLogo from './img/npm.svg';
import pnpmLogo from './img/pnpm.svg';
import yarnLogo from './img/yarn.svg';

const InstallItem = styled(ListItem)({
  padding: 0,
  ':hover': {
    backgroundColor: 'transparent',
  },
});

const InstallListItemText = styled(ListItemText)({
  padding: '0 10px',
  margin: 0,
});

const PackageMangerAvatar = styled(Avatar)({
  borderRadius: '0px',
  padding: '0',
  img: {
    backgroundColor: 'transparent',
  },
});

export enum DependencyManager {
  NPM = 'npm',
  YARN = 'yarn',
  PNPM = 'pnpm',
}

interface Interface {
  packageName: string;
  dependencyManager: DependencyManager;
}

const InstallListItem: React.FC<Interface> = ({ packageName, dependencyManager }) => {
  const { t } = useTranslation();

  switch (dependencyManager) {
    case DependencyManager.NPM:
      return (
        <InstallItem button={true} data-testid={'installListItem-npm'}>
          <PackageMangerAvatar alt="npm" src={npmLogo} />
          <InstallListItemText
            primary={<CopyToClipBoard text={t('sidebar.installation.install-using-npm-command', { packageName })} />}
            secondary={t('sidebar.installation.install-using-npm')}
          />
        </InstallItem>
      );
    case DependencyManager.YARN:
      return (
        <InstallItem button={true} data-testid={'installListItem-yarn'}>
          <PackageMangerAvatar alt="yarn" src={yarnLogo} />
          <InstallListItemText
            primary={<CopyToClipBoard text={t('sidebar.installation.install-using-yarn-command', { packageName })} />}
            secondary={t('sidebar.installation.install-using-yarn')}
          />
        </InstallItem>
      );
    case DependencyManager.PNPM:
      return (
        <InstallItem button={true} data-testid={'installListItem-pnpm'}>
          <PackageMangerAvatar alt={'pnpm'} src={pnpmLogo} />
          <InstallListItemText
            primary={<CopyToClipBoard text={t('sidebar.installation.install-using-pnpm-command', { packageName })} />}
            secondary={t('sidebar.installation.install-using-pnpm')}
          />
        </InstallItem>
      );
    default:
      return null;
  }
};

export default InstallListItem;
