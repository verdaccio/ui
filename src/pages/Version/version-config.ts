import { PackageMetaInterface } from '../../../types/packageMeta';

export interface DetailContextProps {
  enableLoading: () => void;
  hasNotBeenFound: boolean;
  isLoading: boolean;
  packageMeta: PackageMetaInterface;
  packageName: string;
  packageVersion?: string;
  readMe: string;
}

export interface VersionPageConsumerProps {
  enableLoading: () => void;
  packageMeta: PackageMetaInterface;
  packageName: string;
  packageVersion?: string;
  readMe: string;
}
