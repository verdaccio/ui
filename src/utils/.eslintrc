{
    "rules": {
        "no-invalid-this": 0
    }
}