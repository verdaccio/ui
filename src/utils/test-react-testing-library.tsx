import { render } from '@testing-library/react';
import React from 'react';
import { I18nextProvider } from 'react-i18next';

import ThemeProvider from 'verdaccio-ui/design-tokens/ThemeProvider';
import APIProvider from 'verdaccio-ui/providers/API/APIProvider';
import AppConfigurationProvider from 'verdaccio-ui/providers/config';

import i18nConfig from '../../i18n/config';

const customRender = (node: React.ReactElement, ...options: any) => {
  return render(
    <AppConfigurationProvider>
      <APIProvider>
        <ThemeProvider>
          <I18nextProvider i18n={i18nConfig}>{node}</I18nextProvider>
        </ThemeProvider>
      </APIProvider>
    </AppConfigurationProvider>,
    ...options
  );
};

export * from '@testing-library/react';
export { customRender as render };
