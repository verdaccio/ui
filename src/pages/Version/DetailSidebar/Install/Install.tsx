import styled from '@emotion/styled';
import React, { useContext } from 'react';
import { useTranslation } from 'react-i18next';

import List from 'verdaccio-ui/components/List';
import Text from 'verdaccio-ui/components/Text';
import { Theme } from 'verdaccio-ui/design-tokens/theme';

import { DetailContext } from '../..';

import InstallListItem, { DependencyManager } from './InstallListItem';

const StyledText = styled(Text)<{ theme?: Theme }>(props => ({
  fontWeight: props.theme && props.theme.fontWeight.bold,
  textTransform: 'capitalize',
}));

const Install: React.FC = () => {
  const { t } = useTranslation();
  const detailContext = useContext(DetailContext);

  const { packageMeta, packageName } = detailContext;

  if (!packageMeta || !packageName) {
    return null;
  }

  return (
    <List
      data-testid={'installList'}
      subheader={<StyledText variant={'subtitle1'}>{t('sidebar.installation.title')}</StyledText>}>
      <InstallListItem dependencyManager={DependencyManager.NPM} packageName={packageName} />
      <InstallListItem dependencyManager={DependencyManager.YARN} packageName={packageName} />
      <InstallListItem dependencyManager={DependencyManager.PNPM} packageName={packageName} />
    </List>
  );
};

export default Install;
