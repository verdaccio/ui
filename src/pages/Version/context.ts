import { createContext, Consumer, Provider } from 'react';

import { DetailContextProps, VersionPageConsumerProps } from './version-config';

export const DetailContext = createContext<Partial<DetailContextProps>>({});

export const DetailContextProvider: Provider<Partial<VersionPageConsumerProps>> = DetailContext.Provider;
export const DetailContextConsumer: Consumer<Partial<VersionPageConsumerProps>> = DetailContext.Consumer;
