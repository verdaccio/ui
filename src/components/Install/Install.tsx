import React, { useContext } from 'react';
import styled from '@emotion/styled';

import { DetailContext } from '../../pages/Version';
import Text from '../../muiComponents/Text';
import List from '../../muiComponents/List';
import { Theme } from '../../design-tokens/theme';

import InstallListItem, { DependencyManager } from './InstallListItem';

const StyledText = styled(Text)<{ theme?: Theme }>(props => ({
  fontWeight: props.theme && props.theme.fontWeight.bold,
  textTransform: 'capitalize',
}));

const Install: React.FC = () => {
  const detailContext = useContext(DetailContext);

  const { packageMeta, packageName } = detailContext;

  if (!packageMeta || !packageName) {
    return null;
  }

  return (
    <List data-testid={'installList'} subheader={<StyledText variant={'subtitle1'}>{'Installation'}</StyledText>}>
      <InstallListItem dependencyManager={DependencyManager.NPM} packageName={packageName} />
      <InstallListItem dependencyManager={DependencyManager.YARN} packageName={packageName} />
      <InstallListItem dependencyManager={DependencyManager.PNPM} packageName={packageName} />
    </List>
  );
};

export default Install;
