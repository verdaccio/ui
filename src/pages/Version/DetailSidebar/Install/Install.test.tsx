import React from 'react';

import { render } from 'verdaccio-ui/utils/test-react-testing-library';

import { DetailContext } from '../../context';
import { DetailContextProps } from '../../version-config';

import data from './__partials__/data.json';
import Install from './Install';

const detailContextValue: Partial<DetailContextProps> = {
  packageName: 'foo',
  packageMeta: data,
};

const ComponentToBeRendered: React.FC = () => (
  <DetailContext.Provider value={detailContextValue}>
    <Install />
  </DetailContext.Provider>
);

/* eslint-disable react/jsx-no-bind*/
describe('<Install />', () => {
  test('renders correctly', () => {
    const { container } = render(<ComponentToBeRendered />);
    expect(container.firstChild).toMatchSnapshot();
  });

  test('should have 3 children', () => {
    const { getByTestId } = render(<ComponentToBeRendered />);
    const installListItems = getByTestId('installList');
    // installitems + subHeader = 4
    expect(installListItems.children.length).toBe(4);
  });

  test('should have the element NPM', () => {
    const { getByTestId, queryByText } = render(<ComponentToBeRendered />);
    expect(getByTestId('installListItem-npm')).toBeTruthy();
    expect(queryByText(`npm install ${detailContextValue.packageName}`)).toBeTruthy();
    expect(queryByText('Install using npm')).toBeTruthy();
  });

  test('should have the element YARN', () => {
    const { getByTestId, queryByText } = render(<ComponentToBeRendered />);
    expect(getByTestId('installListItem-yarn')).toBeTruthy();
    expect(queryByText(`yarn add ${detailContextValue.packageName}`)).toBeTruthy();
    expect(queryByText('Install using yarn')).toBeTruthy();
  });

  test('should have the element PNPM', () => {
    const { getByTestId, queryByText } = render(<ComponentToBeRendered />);
    expect(getByTestId('installListItem-pnpm')).toBeTruthy();
    expect(queryByText(`pnpm install ${detailContextValue.packageName}`)).toBeTruthy();
    expect(queryByText('Install using pnpm')).toBeTruthy();
  });
});
