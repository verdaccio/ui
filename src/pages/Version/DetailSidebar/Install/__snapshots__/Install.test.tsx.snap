// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Install /> renders correctly 1`] = `
.emotion-0 {
  font-weight: 700;
  text-transform: capitalize;
}

.emotion-2 {
  padding: 0;
}

.emotion-2:hover {
  background-color: transparent;
}

.emotion-4 {
  border-radius: 0px;
  padding: 0;
}

.emotion-4 img {
  background-color: transparent;
}

.emotion-6 {
  padding: 0 10px;
  margin: 0;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.emotion-10 {
  display: inline-block;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  height: 21px;
  font-size: 1rem;
}

<ul
  class="MuiList-root MuiList-padding MuiList-subheader"
  data-testid="installList"
>
  <span
    class="MuiTypography-root emotion-0 emotion-1 MuiTypography-subtitle1"
  >
    Installation
  </span>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
    data-testid="installListItem-npm"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiAvatar-root MuiAvatar-circle emotion-4 emotion-5"
    >
      <img
        alt="npm"
        class="MuiAvatar-img"
        src="[object Object]"
      />
    </div>
    <div
      class="MuiListItemText-root emotion-6 emotion-7 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        <div
          class="emotion-8 emotion-9"
        >
          <span
            class="emotion-10 emotion-11"
          >
            npm install foo
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            data-testid="copy-icon"
            tabindex="0"
            title="Copy to clipboard"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4l6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2h7zm-1 7h5.5L14 6.5V12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        Install using npm
      </p>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
    data-testid="installListItem-yarn"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiAvatar-root MuiAvatar-circle emotion-4 emotion-5"
    >
      <img
        alt="yarn"
        class="MuiAvatar-img"
        src="[object Object]"
      />
    </div>
    <div
      class="MuiListItemText-root emotion-6 emotion-7 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        <div
          class="emotion-8 emotion-9"
        >
          <span
            class="emotion-10 emotion-11"
          >
            yarn add foo
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            data-testid="copy-icon"
            tabindex="0"
            title="Copy to clipboard"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4l6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2h7zm-1 7h5.5L14 6.5V12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        Install using yarn
      </p>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
  <div
    aria-disabled="false"
    class="MuiButtonBase-root MuiListItem-root emotion-2 emotion-3 MuiListItem-gutters MuiListItem-button"
    data-testid="installListItem-pnpm"
    role="button"
    tabindex="0"
  >
    <div
      class="MuiAvatar-root MuiAvatar-circle emotion-4 emotion-5"
    >
      <img
        alt="pnpm"
        class="MuiAvatar-img"
        src="[object Object]"
      />
    </div>
    <div
      class="MuiListItemText-root emotion-6 emotion-7 MuiListItemText-multiline"
    >
      <span
        class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
      >
        <div
          class="emotion-8 emotion-9"
        >
          <span
            class="emotion-10 emotion-11"
          >
            pnpm install foo
          </span>
          <button
            class="MuiButtonBase-root MuiIconButton-root"
            data-testid="copy-icon"
            tabindex="0"
            title="Copy to clipboard"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M16 1H4c-1.1 0-2 .9-2 2v14h2V3h12V1zm-1 4l6 6v10c0 1.1-.9 2-2 2H7.99C6.89 23 6 22.1 6 21l.01-14c0-1.1.89-2 1.99-2h7zm-1 7h5.5L14 6.5V12z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </span>
      <p
        class="MuiTypography-root MuiListItemText-secondary MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-displayBlock"
      >
        Install using pnpm
      </p>
    </div>
    <span
      class="MuiTouchRipple-root"
    />
  </div>
</ul>
`;
